From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Mon, 18 Jun 2018 01:09:27 -0400
Subject: [PATCH] PlayerReadyArrowEvent and PlayerReadyProjectileEvent

Called when a player is firing a bow/crossbow and the server is choosing an projectile to use.
Plugins can skip selection of certain projectiles and control which is used.

Co-authored-by: Doc <nachito94@msn.com>

diff --git a/src/main/java/com/destroystokyo/paper/event/player/PlayerReadyArrowEvent.java b/src/main/java/com/destroystokyo/paper/event/player/PlayerReadyArrowEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..2aca81b46dae564bc18a117668f6d3aa57ff99c1
--- /dev/null
+++ b/src/main/java/com/destroystokyo/paper/event/player/PlayerReadyArrowEvent.java
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2018 Daniel Ennis (Aikar) MIT License
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining
+ *  a copy of this software and associated documentation files (the
+ *  "Software"), to deal in the Software without restriction, including
+ *  without limitation the rights to use, copy, modify, merge, publish,
+ *  distribute, sublicense, and/or sell copies of the Software, and to
+ *  permit persons to whom the Software is furnished to do so, subject to
+ *  the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package com.destroystokyo.paper.event.player;
+
+import io.papermc.paper.event.player.PlayerReadyProjectileEvent;
+import org.bukkit.entity.Player;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Called when a player is firing a bow or crossbow and the server is choosing an arrow to use.
+ *
+ * @deprecated Deprecated in favor of {@link PlayerReadyProjectileEvent}
+ */
+@Deprecated
+public class PlayerReadyArrowEvent extends PlayerReadyProjectileEvent {
+
+    public PlayerReadyArrowEvent(@NotNull Player player, @NotNull ItemStack bow, @NotNull ItemStack arrow) {
+        super(player, bow, arrow);
+    }
+
+    /**
+     * @return the player is using to fire the arrow
+     */
+    @NotNull
+    public ItemStack getBow() {
+        return super.getWeaponItem();
+    }
+
+    /**
+     * @return the arrow that is attempting to be used
+     */
+    @NotNull
+    public ItemStack getArrow() {
+        return super.getProjectileItem();
+    }
+
+    /**
+     * Whether or not use of this arrow is cancelled. On cancel, the server will try the next arrow available and fire another event.
+     */
+    @Override
+    public boolean isCancelled() {
+        return super.isCancelled();
+    }
+
+    /**
+     * Cancel use of this arrow. On cancel, the server will try the next arrow available and fire another event.
+     * @param cancel true if you wish to cancel this event
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        super.setCancelled(cancel);
+    }
+}
diff --git a/src/main/java/io/papermc/paper/event/player/PlayerReadyProjectileEvent.java b/src/main/java/io/papermc/paper/event/player/PlayerReadyProjectileEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..2db0bd008f30aa2181a932325520661dbf70278b
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/player/PlayerReadyProjectileEvent.java
@@ -0,0 +1,64 @@
+package io.papermc.paper.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerReadyProjectileEvent extends PlayerEvent implements Cancellable {
+
+    private static final HandlerList handlers = new HandlerList();
+    @NotNull private final ItemStack projectileWeapon;
+    @NotNull private final ItemStack projectileMunition;
+    private boolean cancelled = false;
+
+    public PlayerReadyProjectileEvent(@NotNull Player player, @NotNull ItemStack projectileWeapon, @NotNull ItemStack projectileMunition) {
+        super(player);
+        this.projectileWeapon = projectileWeapon;
+        this.projectileMunition = projectileMunition;
+    }
+
+    /**
+     * @return the weapon is using to fire the projectile
+     */
+    @NotNull
+    public ItemStack getWeaponItem() {
+        return this.projectileWeapon;
+    }
+
+    /**
+     * @return the projectile that is attempting to be used
+     */
+    @NotNull
+    public ItemStack getProjectileItem() {
+        return this.projectileMunition;
+    }
+
+    /**
+     * Whether or not use of this arrow is cancelled. On cancel, the server will try the next arrow available and fire another event.
+     */
+    @Override
+    public boolean isCancelled() {
+        return this.cancelled;
+    }
+
+    /**
+     * Cancel use of this projectile. On cancel, the server will try the next projectile available in base of the {@link #getWeapon()} and fire another event.
+     * @param cancel true if you wish to cancel this event
+     */
+    @Override
+    public void setCancelled(boolean cancel) {
+        this.cancelled = cancel;
+    }
+
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+}
